src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/AutoInsert/RemoteAutoInsertService.cs

```diff
@@ -0,0 +1,218 @@
+        CancellationToken cancellationToken)
+    {
+        // Special case for C# where we use AutoInsert for two purposes:
+        // 1. For XML documentation comments (filling out the template when typing "///")
+        // 2. For "on type formatting" style behavior, like adjusting indentation when pressing Enter inside empty braces
+        //
+        // If users have turned off on-type formatting, they don't want the behavior of number 2, but its impossible to separate
+        // that out from number 1. Typing "///" could just as easily adjust indentation on some unrelated code higher up in the
+        // file, which is exactly the behavior users complain about.
+        //
+        // Therefore we are just going to no-op if the user has turned off on type formatting. Maybe one day we can make this
+        // smarter, but at least the user can always turn the setting back on, type their "///", and turn it back off, without
+        // having to restart VS. Not the worst compromise (hopefully!)
+        if (!formatOnType)
+        {
+            return Response.NoFurtherHandling;
+        }
+
+        if (!AutoInsertService.CSharpAllowedAutoInsertTriggerCharacters.Contains(character))
+        {
+            return Response.NoFurtherHandling;
+        }
+
+        var generatedDocument = await remoteDocumentContext.Snapshot.GetGeneratedDocumentAsync().ConfigureAwait(false);
+        var formattingOptions = new RoslynFormattingOptions()
+        {
+            InsertSpaces = !indentWithTabs,
+            TabSize = indentSize
+        };
+        var autoInsertResponseItem = await OnAutoInsert.GetOnAutoInsertResponseAsync(
+            generatedDocument,
+            mappedPosition,
+            character,
+            formattingOptions,
+            cancellationToken
+        );
+
+        if (autoInsertResponseItem is null)
+        {
+            return Response.NoFurtherHandling;
+        }
+
+        var razorFormattingOptions = new VsLspFormattingOptions()
+        {
+            InsertSpaces = !indentWithTabs,
+            TabSize = indentSize
+        };
+
+        var mappedEdits = autoInsertResponseItem.TextEditFormat == RoslynInsertTextFormat.Snippet
+            ? await _razorFormattingService.FormatSnippetAsync(
+                remoteDocumentContext,
+                RazorLanguageKind.CSharp,
+                [autoInsertResponseItem.TextEdit.ToVsLspTextEdit()],
+                razorFormattingOptions,
+                cancellationToken)
+            .ConfigureAwait(false)
+            : await _razorFormattingService.FormatOnTypeAsync(
+                remoteDocumentContext,
+                RazorLanguageKind.CSharp,
+                [autoInsertResponseItem.TextEdit.ToVsLspTextEdit()],
+                razorFormattingOptions,
+                hostDocumentIndex: 0,
+                triggerCharacter: '\0',
+                cancellationToken)
+            .ConfigureAwait(false);
+
+        if (mappedEdits is not [{ } edit])
```

Comment:
Is it an error condition that `mappedEdits` contains just a single element? If so, should we `Debug.Fail(...)` or something like that?

Reply:
This was the existing code I moved. I was going to follow up, but since David already addressed this I am going to look at what he did instead I guess. I'll resolved it in this PR.